[BE] Refresh documentation for stable ABI / API (#163899)
pytorch/pytorch: docs/source/notes/libtorch_stable_abi.md

# LibTorch Stable ABI

## Overview

> The LibTorch Stable ABI (Application Binary Interface) provides a stable interface for extending PyTorch functionality without being tightly coupled to specific PyTorch versions. This enables the development of custom operators and extensions that remain compatible across PyTorch releases.

Reviewer: ```suggestion
The LibTorch Stable ABI (Application Binary Interface) provides an interface for extending PyTorch functionality without being tightly coupled to specific PyTorch versions. This enables the development of custom operators and extensions that remain compatible across PyTorch releases.
```
nit

The stable ABI consists of three main components:

1. **Stable C headers** - Low-level C API implemented by libtorch (primarily `torch/csrc/inductor/aoti_torch/c/shim.h`)

> 2. **Header-only C++ library** - Standalone utilities (`torch/headeronly/*`)

Reviewer: ```suggestion
2. **Header-only C++ library** - Standalone utilities implemented in only headers such that there is no dependence on libtorch (`torch/headeronly/*`)
```

3. **Stable C++ wrappers** - High-level C++ convenience wrappers (`torch/csrc/stable/*`)

We discuss each of these in detail

### Stable C headers

> The stable C headers used by AOTInductor form the foundation of the stable ABI. However, this is **use at your own risk**. For example, users must handle the memory lifecycle of objects returned by certain APIs.
>  Further, the stack-based APIs discussed below which allow the user to call the PyTorch dispatcher don't provide strong guarantees on forward and backward compatibility.
> 
> Unless absolutely necessary, we recommend the high-level C++ API in `torch/csrc/stable`
> which will handle all the rough edges of the C API for the user.

Reviewer: ```suggestion
The stable C headers form the foundation of the stable ABI and are not meant to be user facing. These C-style APIs are ABI stable and do the work of liaisoning with the underlying libtorch.so implementations. You may **use at your own risk**. For example, you must handle the memory lifecycle of objects returned by certain APIs if you decide to use them! We've intentionally designed the C++ APIs mentioned below in `torch/headeronly` or `torch/csrc/stable` to be significantly more user friendly. Unless absolutely necessary, we recommend using these high-level C++ APIs
which will handle all the rough edges of the C API for you.

Furthermore, if you do decide to embark and use the stack-based APIs discussed below which allow you to call the PyTorch dispatcher, note that they do not provide strong guarantees on forward and backward compatibility of the underlying op.
```

maybe we should rearrange mentioning this to third as we don't want people to use these directly

### `torch/headeronly`

This is a set of inlined C++ headers are completely decoupled from libtorch. The headers consist of certain utilities that might be familiar to custom extension writers. For example, the
`c10::ScalarType` enum lives here as `torch::headeronly::ScalarType`.

### `torch/csrc/stable`

> This is a set of inlined C++ headers that provide wrappers around the C API that handle the footguns

Reviewer: ```suggestion
This is a set of inlined C++ headers that provide wrappers around the C API that handle the rough edges
```

discussed above.

It consists of

- torch/csrc/stable/library.h: Provides a stable version of TORCH_LIBRARY and similar macros.

> - torch/csrc/stable/tensor.h: Provides torch::stable::Tensor, a stable version of at::Tensor.

Reviewer: ```suggestion
- torch/csrc/stable/tensor_struct.h: Provides torch::stable::Tensor, a stable version of at::Tensor.
```

> - torch/csrc/stable/ops.h: Provides a stable interface for calling ops from `native_functions.yaml`.

Reviewer: ```suggestion
- torch/csrc/stable/ops.h: Provides a stable interface for calling ATen ops from `native_functions.yaml`.
```

- torch/csrc/stable/accelerator.h: Provides a stable interface for device-generic objects and APIs

> (e.g. `getCurrentStream`).

Reviewer: also mention DeviceGuard!

> The interface provided by `torch/csrc/stable` should be familiar to custom extension writers but might not match perfectly in certain cases.

Reviewer: ```suggestion
We are continuing improving coverage in our `torch/csrc/stable` APIs. Please file an issue if you'd like to see support for particular APIs in your custom extension.
```

> ## How are objects passed across ABI boundaries when interacting with the dispatcher?

Reviewer: ```suggestion
## How are objects passed across the ABI boundary when interacting with the dispatcher?
```

2. The stack always has ownership of the objects it holds.
    a. When calling a stack-based API, you must give owning references to the calling stack and steal references from the returned stack.
    b. When registering your function to be called with a stack, you must steal references from your argument stack and push onto the stack new references.

### Stack-based APIs

The above is relevant in two places:

1. `STABLE_TORCH_LIBRARY`

>     Unlike `TORCH_LIBRARY`, the dispatcher expects kernels registered via `STABLE_TORCH_LIBRARY` to be boxed. This means they must have the signature `(StableIValue* stack, uint64_t num_args, uint64_t num_outputs) -> void`. This will soon be abstracted away from the user, but for the time being, users can use the `from` and `to`.

Reviewer: ```suggestion
    Unlike `TORCH_LIBRARY`, the dispatcher expects kernels registered via `STABLE_TORCH_LIBRARY` to be boxed. This means they must have the signature `(StableIValue* stack, uint64_t num_args, uint64_t num_outputs) -> void`. We plan to eventually abstract away the need for manual boxing, but, for the time being, please use `from` and `to`.
```

    ```cpp
    Tensor my_amax_vec(Tensor t) {
        std::vector<int64_t> v = {0,1};
        return amax(t, v, false);
    }

    void boxed_my_amax_vec(StableIValue* stack, uint64_t num_args, uint64_t num_outputs) {
        auto res = my_amax_vec(to<Tensor>(stack[0]));
        stack[0] = from(res);
    }
    ```

> 2. `aoti_torch_call_dispatcher`

Reviewer: Can you mention what it does too? Feel free to copy the docs I wrote in-code.
